presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/trait/CardinalityGroupTraitCalculator.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative.trait;
+
+import com.facebook.presto.sql.planner.iterative.GroupReference;
+import com.facebook.presto.sql.planner.iterative.GroupTraitCalculator;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.LimitNode;
+import com.facebook.presto.sql.planner.plan.MarkDistinctNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.google.common.collect.Range;
+
+import java.util.List;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.Math.min;
+
+public final class CardinalityGroupTraitCalculator
+        implements GroupTraitCalculator<CardinalityGroupTrait>
+{
+    private final Visitor visitor = new Visitor();
+
+    @Override
+    public CardinalityGroupTrait calculate(PlanNode node, List<CardinalityGroupTrait> sourceTraits)
+    {
+        return node.accept(visitor, sourceTraits);
+    }
+
+    private static final class Visitor
+            extends PlanVisitor<CardinalityGroupTrait, List<CardinalityGroupTrait>>
```

Comment:
Yeah, I want to get rid of the Visitor interface for PlanNodes. It's one of the things preventing us from having an extensible (i.e., via plugins) set of nodes and rules.

Instead, we should take one of these approaches:
- Have each node (optionally) compute the trait based on the traits of its inputs (e.g., something like `ProjectNode.computeCardinality(List<CardinalityTrait> inputCardinalities)`
- Have some form of registry where node types and trait handlers are registered and looked up. This is probably overkill for now, but may be necessary when we want to support plugins, many more traits, etc.

Both of the above are similar in essence, so we can always refactor and adapt them if needed.